includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -408,17 +409,21 @@ public function payment_fields() {
 	/**
 	 * Process the payment for a given order.
 	 *
-	 * @param int  $order_id Order ID to process the payment for.
-	 * @param bool $force_save_payment_method Whether this is a one-off payment (false) or it's the first installment of a recurring payment (true).
+	 * @param int    $order_id                  Order ID to process the payment for.
+	 * @param bool   $force_save_payment_method Whether to save the payment method upon successful payment.
+	 * @param string $payment_type              The type of the payment that's being made.
 	 *
 	 * @return array|null An array with result of payment and redirect URL, or nothing.
 	 */
-	public function process_payment( $order_id, $force_save_payment_method = false ) {
+	public function process_payment( $order_id, $force_save_payment_method = false, $payment_type = null ) {
```

Comment:
Adding one more "internal" parameter to the public method doesn't look right to me. It makes things pretty complex with multiple calls just passing those args around. I was thinking quite long about this method and how we can avoid passing those params around, so it looks like we need to remove both `$force_save_payment_method` and `$payment_type` because they are internal and used only by subscriptions [here](https://github.com/Automattic/woocommerce-payments/blob/fix%2Fdecouple-force-save-from-metadata/includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php#L91).

The idea is the following:
1. Final payment processing is made in `process_payment_for_order`, while process_payment only prepares payment_information instance. So `WC_Payment_Gateway_WCPay` and `WC_Payment_Gateway_WCPay_Subscriptions_Compat` should do it themselves.
2. Process payment for order should not accept `$force_save_payment_method` as an argument. It should rely on new `$payment_information->should_save_payment_method()`.
3. Instead of falling back to `parrent::process_payment(...)` [here](https://github.com/Automattic/woocommerce-payments/blob/fix%2Fdecouple-force-save-from-metadata/includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php#L91) we could make appropriate `payment_information` instance and call `$this->process_order_for_payment(...)` directly.

What do you think? 

cc @reykjalin

Reply:
As we discussed in Slack PMs, I:

- Isolated the generation of `Payment_Information` in a separate `prepare_payment_information` method, which gets is easy to overload in the subs class.
- As the next step after `prepare_payment_information`, the flat that indicates whether the PM should be saved got moved to `Payment_Information` and is set independently in each class, without having to overwrite/force it.
- Now `process_payment` and `prepare_payment_information` only accept the order (id) as an argument and theother flags/parameters are handled much more elegantly.